### ISSUE TITLE
Is tracker-react production ready? Replace ReactMeteorData

### ISSUE THREAD

**Issue by D1no** [ultimatejs/tracker-react#1]
Hey James,
first of all, I want to thank you for your relentless engagement on the react<->meteor integration and the blaze/sideburnz/tracker-react front. Tracking along here and there, I was amazed by some of the genuine solutions you came up with.

**tracker-react**
I would like to start using it as a drop in replacement in react for the current [ReactMeteorData mixin](https://react-in-meteor.readthedocs.org/en/latest/meteor-data/#the-reactmeteordata-mixin). I have done some tests, talked with my engineers, and it seems to be solid (except for missing tests).

**status**
Is tracker-react a viable solution for ReactMeteorData? @tmeasday I know you took a quick look at it - any resume or MDG statement about interest in aligning the two mixins?

At some point, we are ultimately interested in proxying everything through [redux](http://redux.js.org/) to have full control over the application state to ensure predictability and enable more sophisticated debugging tools (while maintaing flexebility). Currently, our stack looks similar to the [meteor webpack kistart-hugeapp](https://github.com/thereactivestack/kickstart-hugeapp) repo + typescript.

PS:
I know you are very active - i.e. on the forums - but currently all the discussions are scattered all over the place... which makes it hard to follow. Would be great if you could collect the discussion links for the specific projects and note them in their github readme.md's. This way the github repos are true entry points to follow for more info.

**Comment by faceyspacey** [comment]
@D1no

> but currently all the discussions are scattered all over the place... which makes it hard to follow. Would be great if you could collect the discussion links for the specific projects and note them in their github readme.md's. 

I got a few more mixins I'm about to release. Then I'm going to consolidate them all and release `Meteor.Component` as a drop-in replacement for `React.Component`, eg:

```
class MyComponent extends Meteor.Component {
   render() {

   }
}
```

When I get to that point I'm going to start consolidating all the info. Right now my goal is to get the code done, and get it to be just more than a bunch of scattered potentially outdated or not fully-functioning mixins. So yea, it will be documented very well, and my ultimate goal is to get MDG to just adopt it as their Meteor component system (Blaze 2, whatever you wanna call it). 

ps. Glad you like it. Let the community know in the forum and crater of any successful experiments you do with it. This is just the beginning.

**Comment by D1no** [comment]
hey @faceyspacey thanks for your quick reply. Do you have any ETA on your `Meteor.Component` approach? From my view, I am fine with a mixin.

(I also hope you are not short-betting yourself on the blaze-react transpiler. MGD has historically been resilient to pick up contributions... from my believe, blaze and react solve two different things and should therefore rather co-exist. tracker<->react<->redux integrations are a real pain right now that hinder full meteor potential. Stuck in the middle.)

**Comment by faceyspacey** [comment]
"Short betting myself" lol.

It's hard to say. Mainly it's that @timbrandin has been doing the spacebars transpilation stuff and I realized a while back if we couldn't replicate blaze behavior, it wouldn't matter if we could transpile spacebars. So that's why I've been focusing on bringing blaze behavior to React _first_.

That said i do have a hunch these things will be even more popular than blaze 1 transpilation, especially if there are any kinks with the blaze 1 transpilation. If it can't be perfect it just doesn't work, u know. My backup strategy if that ends up being the case is a _one time transpilation step that overwrites your files_ which helps u upgrade ur code to the new Meteor.Component system. So in that case u run the transpiler once and write the changes to the file system, and then basically a linter of sorts (possibly run by a GUI) will tell you other changes you gotta manually make and recommendations for your options. 

But things are looking extremely good. I'm just trying to prepare for the worst in addition.

Ps. I got some things in mind for redux too. I haven't put too much thought into it--perhaps u have a better idea--but basically:

_1. we use a_ **_TrackerReact**_-_inspired mechanism to trigger actions in Redux when a reactive data dependency is invalidated
2. then automatically reduce reactive data sources into the next state
3. then invoke all listeners attached to components._ 

You may ask what the point is when Tracker + MiniMongo basically does the job of most of Redux? The point is so to provide a consistent redux way of building things. There are plenty of things where you won't use mini mongo and certainly not want to use the global session var. You may want to take advantage of Redux's time traveling debugger, etc. This just funnels reactive data sources through redux before reaching components before hitting components directly. It's just an idea currently, but once we build a little of it the rest of the terrain will become visible. 

There's a lot of talk about using the Flux pattern server-side as well, i.e. "over the wire":
https://blog.rotenberg.io/flux-over-the-wire-3/

Couple that with being a _subscription_ and you have a lot of what GraphQL is trying to achieve with its upcoming "subscriptions" specification: 

https://gist.github.com/OlegIlyenko/a5a9ab1b000ba0b5b1ad

Which ultimately looks almost identical to Meteor subscriptions! Minus the side-effect_less_ immutable+functional+unidirectional benefits, plus the fact LiveQuery is tailing a growing query on every server in your cluster (GraphQL subscriptions would happen on the web server layer before the DB is hit, i.e. pre-writes).

In short, we have 3 options:
1. current Meteor subscriptions/publications (based on LiveQuery)
2. something built in the future on top of GraphQL subscriptions (in the web server layer)
3. perhaps something TODAY based on _"Redux Over the Wire"_

#3 may in fact be quite easy to build with Meteor. I think the approach would be a combination of Redux over the wire, plus reactive subscriptions (i.e. observed queries) over the wire. 

At the end of the day, i think the most important thing is we figure out **WHY** we wanna do any of this? Just to use Redux, the newest coolest tool in the box? _OR_, for specific things that Redux gives us. The idea being I imagine a sort of combined meteor style reactivity + redux unidirectional reactivity approach. So in whatever we do, we cant stomp out the best of Redux with Meteor's approach. Rather, we need to highlight the best of Redux, while bringing some of the automation (and deep Mongo integration) Meteor already has into the fold. 

So what's the best of Redux? **Time traveling? No side effects? Decoupling** **_shared state**_ **from multiple components that may rely on it? Anything else?**

In short, my guess is we can accomplish all 3, but with minor side effects, which we can hopefully insulate the user from in most scenarios :). Let me know if you have any other goals here. We should build this.

**Comment by ghost** [comment]
Hi @faceyspacey :)

it's been a while, and wanted to ask if you're making progress with Tracker-React? I'd really love to use React given full tracker support :)

**Comment by faceyspacey** [comment]
There isn't much really to do with it. It's being incorporated into a larger project. But tracker react is and has been pretty much solid as is. It's not a big library. View the code and u can easily see the mechanism by which it works. If we ever come across any issues, which we haven't yet, I can likely fix them quickly. 

The main thing is tracker needs to be made to be more like Mobservable so there are no (or less) unnecessary re-runs. I've been busy with other stuff so I haven't gotten a chance to do that or think more about it. A breakthrough has to occur there in terms of how to upgrade tracker and I haven't hit that point yet. I've mainly been studying everything else out there to broaden my perspective. 

Anyway try Tracker React out. It will be quite pleasant and surprising how u can get the best of both worlds: functional props based unidirectional rendering and the ease of ad hoc automatic + transparent reactivity.

**Comment by ghost** [comment]
I've used it quite a bit already - it's very nice. I've had some problems though - but that's probably on me then :) Will look deeper into it. 

Yea, it surely is a really simple solution! :+1: 

Been looking into Elm as well, but FRP and functional programming in general is very different from the basics I've been taught so far :) fullstack elm would truely be cool ^^ or even better - isomorphic fullstack elm xD elm:signals are awesome!

**Comment by D1no** [comment]
We completely switched out GetMeteorData for TrackerReact. Just make sure to subscribe to [publications on the constructor level (https://github.com/D1no/touch2s/blob/master/modules/TodoApp/client/TodoMain.jsx#L19-L31), so they server side render.

But the best side effect: Quick and dirty prototypes are back! If you missed some data, just throw it happily into the deepest button component. Just like in the old blaze hacking days ! ... though James probably does not like to hear that :b

**Comment by Goatic** [comment]
haha :D I've been subscribing within the "componentDidMount" hook, as I figured it would only run once xD obviously, I don't know React well enough... will make sure to move them to the constructer, and I figure my worries will be over :+1:  thnx

**Comment by faceyspacey** [comment]
@D1no is there something wrong with subscribing to publications in `componentWillMount` on the server?

**Comment by D1no** [comment]
No, in react ES6 `constructor` === `componentWillMount`. But what does not work is to initialise i.e. as default props. Also, `componentDidMount` seems to be ignored in some odd cases. (<- Remember, talking about Server Side Rendering)
